lib/KTable/index.vue

```diff
@@ -0,0 +1,484 @@
+        SORT_ORDER_ASC,
+        SORT_ORDER_DESC,
+        DATA_TYPE_OTHERS,
+        getHeaderStyle,
+      };
+    },
+    /* eslint-disable kolibri/vue-no-unused-properties */
+    props: {
+      /**
+       * An array of objects `{ label, dataType, minWidth, width }`representing the headers of the table. The `dataType` can be one of `'string'`, `'numeric'`, `'date'`, or `'others'`. `label` and `dataType` are required. `minWidth` and `width` are optional.
+       */
+      headers: {
+        type: Array,
+        required: true,
+        validator: function(value) {
+          return value.every(
+            header =>
+              ['label', 'dataType'].every(key => key in header) &&
+              ['string', 'numeric', 'date', 'others'].includes(header.dataType)
+          );
+        },
+      },
+      /**
+       * An array of arrays representing the rows of the table. Each row should have the same number of elements as the headers array.
+       */
+      rows: {
+        type: Array,
+        required: true,
+      },
+      /**
+       * The caption of the table
+       */
+      caption: {
+        type: String,
+        required: true,
+      },
+      /**
+       * Enables or disables local sorting of the table data.
+       */
+      useLocalSorting: {
+        type: Boolean,
+        default: false,
+      },
+      /**
+       * Enables or disables sorting functionality for the table headers.
+       */
+      sortable: {
+        type: Boolean,
+        default: true,
+      },
+    },
+    data() {
+      return {
+        focusedRowIndex: null,
+        focusedColIndex: null,
+        hoveredRowIndex: null,
+      };
+    },
+    computed: {
+      coreOutlineFocus() {
+        return {
+          ':focus': {
+            ...this.$coreOutline,
+            outlineOffset: '-2px',
+          },
+        };
+      },
+    },
+    methods: {
+      /**
+       * Takes care of
+       *  - keyboard navigation focus trap
+       *  - the arrow keys navigation flow
+       *  - the tab keys navigation flow
+       *  - access to focusable elements within a cell via tab and shift tab keys
+       *  - triggering sort on the enter key
+       *  - header highlight
+       */
+      handleKeydown(event, rowIndex, colIndex) {
+        const key = event.key;
+        const totalRows = this.rows.length;
+        const totalCols = this.headers.length;
+
+        let nextRowIndex = rowIndex;
+        let nextColIndex = colIndex;
+        // Identify all focusable elements inside the current cell
+        const currentCell = this.getCell(rowIndex, colIndex);
+        const focusableElements = currentCell
+          ? Array.from(
+              currentCell.querySelectorAll(
+                'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
+              )
+            )
+          : [];
+        const focusedElementIndex = focusableElements.indexOf(document.activeElement);
+        switch (key) {
```

Comment:
[Optional] I think this could be broken down in smaller handler methods for each key case, and explain a little bit the logic in each of them. For example just reading the code I didnt realize that if you where at the beginning of the row and press left the focus changes to the last item in the previous row. We could have perhaps a handleArrowUp() that returns [nextRowIndex, nextColumnIndex] and just do something like
```javascript
case 'ArrowUp':
  [nextRowIndex, nextColIndex] = this.handleArrowUp();
case 'ArrowDown':
  [nextRowIndex, nextColIndex] = this.handleArrowDown();
```
and so on.

And its even more important with the tab, as it has an even more complex logic.

Reply:
Just a heads up on this one, I tried refactoring the code to the suggested format. All navigations work perfect except the tab key navigation. The code is same as the one that I have been using so far and am unable to figure out what might have gone wrong. Attaching a snippet of that code here (haven't pushed it yet). Are you able to spot anything that might be an issue?
```
handleKeydown(event, rowIndex, colIndex) {
        const key = event.key;
        const totalRows = this.rows.length;
        const totalCols = this.headers.length;

        let nextRowIndex = rowIndex;
        let nextColIndex = colIndex;

        switch (key) {
          case 'ArrowUp':
            [nextRowIndex, nextColIndex] = this.handleArrowUp(rowIndex, colIndex, totalRows);
            break;
          case 'ArrowDown':
            [nextRowIndex, nextColIndex] = this.handleArrowDown(rowIndex, colIndex, totalRows);
            break;
          case 'ArrowLeft':
            [nextRowIndex, nextColIndex] = this.handleArrowLeft(
              rowIndex,
              colIndex,
              totalRows,
              totalCols
            );
            break;
          case 'ArrowRight':
            [nextRowIndex, nextColIndex] = this.handleArrowRight(
              rowIndex,
              colIndex,
              totalRows,
              totalCols
            );
            break;
          case 'Enter':
            if (rowIndex === -1 && this.sortable) {
              this.handleSort(colIndex);
            }
            break;
          case 'Tab':
            [nextRowIndex, nextColIndex] = this.handleTab(
              event,
              rowIndex,
              colIndex,
              totalRows,
              totalCols
            );
            break;
          default:
            return;
        }

        this.focusCell(nextRowIndex, nextColIndex);
        this.focusedRowIndex = nextRowIndex === -1 ? null : nextRowIndex;
        this.focusedColIndex = nextColIndex;

        this.highlightHeader(nextColIndex);

        event.preventDefault();
      },
      
      handleTab(event, rowIndex, colIndex, totalRows, totalCols) {
        let nextRowIndex = rowIndex;
        let nextColIndex = colIndex;
        const currentCell = this.getCell(rowIndex, colIndex);
        const focusableElements = [];

        if (currentCell) {
          const buttons = currentCell.getElementsByTagName('button');
          const links = currentCell.getElementsByTagName('a');
          const inputs = currentCell.getElementsByTagName('input');
          const selects = currentCell.getElementsByTagName('select');
          const textareas = currentCell.getElementsByTagName('textarea');

          focusableElements.push(...buttons, ...links, ...inputs, ...selects, ...textareas);
        }
        console.log('Current cell:', currentCell);
        console.log('Focusable elements:', focusableElements);
        const focusedElementIndex = focusableElements.indexOf(document.activeElement);
        if (focusableElements.length > 0) {
          if (!event.shiftKey) {
            if (focusedElementIndex < focusableElements.length - 1) {
              focusableElements[focusedElementIndex + 1].focus();
              event.preventDefault();
              return [rowIndex, colIndex];
            } else {
              if (colIndex < totalCols - 1) {
                nextColIndex = colIndex + 1;
              } else if (rowIndex < totalRows - 1) {
                nextColIndex = 0;
                nextRowIndex = rowIndex + 1;
              } else {
                return [rowIndex, colIndex];
              }
            }
          } else {
            if (focusedElementIndex < focusableElements.length-1) {
              focusableElements[focusedElementIndex + 1].focus();
              event.preventDefault();
              return [rowIndex, colIndex];
            } else {
              if (colIndex > 0) {
                nextColIndex = colIndex - 1;
              } else if (rowIndex > 0) {
                nextColIndex = totalCols - 1;
                nextRowIndex = rowIndex - 1;
              } else {
                return [rowIndex, colIndex]; // Allow default tab behavior when reaching the last cell
              }
            }
          }
        } else {
          if (!event.shiftKey) {
            if (colIndex < totalCols - 1) {
              nextColIndex = colIndex + 1;
            } else if (rowIndex < totalRows - 1) {
              nextColIndex = 0;
              nextRowIndex = rowIndex + 1;
            } else {
              return [rowIndex, colIndex];
            }
          } else {
            if (colIndex > 0) {
              nextColIndex = colIndex - 1;
            } else if (rowIndex > 0) {
              nextColIndex = totalCols - 1;
              nextRowIndex = rowIndex - 1;
            } else {
              return [rowIndex, colIndex];
            }
          }
        }

        return [nextRowIndex, nextColIndex];
      },
```

Reply:
I will have a look soon :)

Reply:
Nothing immediately obvious, I think I'd need to see this in a working branch

Reply:
https://github.com/BabyElias/kolibri-design-system/tree/code-refactor-trial
The trial branch. @MisRob

Reply:
@BabyElias One of the issues with TAB I observe is that it gets stuck on a cell that has focusable element. This is caused by a difference in how the 'after refactor' version works, even though code looks quite the same.

I think the most clear way to point you I could think of would be to recommend you do

```
console.log('will focus cell', nextRowIndex, nextColIndex);
this.focusCell(nextRowIndex, nextColIndex);
```

and see what happens in both versions.

Before, `focusCell` wasn't called right before focus goes to the button in the cell.
After, `focusCell` is called every time, causing some issues later on that can be observed at various places in code.

Looking at the code, I wonder if `return` statements could be a culprit. My guess would be that this is just one problematic place I spotted.

Here, I'd recommend to not transfer the code 1:1 during this refactor, but focus line by line what needs to happen from the user point of view, and how that translates to the new structure.

Reply:
Okay! I'll give it another try today. In case that does not work up, we'll add it to the follow up issues :)

Reply:
Tried today, still to no avail. It's something up with the return statements for sure. But what is that, I have still not been able to figure out. Tried a few different combinations with those statements but no results as such. Think it might be better to add it to post GSoC task-list :(

Reply:
Yes, no problem :) Better to take some time to think through it

Reply:
Hey, just unresolved to say the reason why the tab doesnt work in those cases as I imagined it could happen when I suggested the refactor :sweat_smile:. (@BabyElias You dont need to read all this comment right now, its still optional and it can be solved in a follow-up issue without problems and you can then revisit it).

Right now the return statements cuts the execution of the `handleKeydown` method, for example here:

```javascript
if (focusedElementIndex < focusableElements.length - 1) {
  focusableElements[focusedElementIndex + 1].focus();
  event.preventDefault();
  return;
}
```
This means that if `focusedElementIndex < focusableElements.length - 1` none of the code below will be executed because we are already focusing the element that needs to be focused and nothing more needs to be done (we dont need the `nextColIndex`, `nextRowIndex`, nor the focusCell method that is called below, etc). 

But in the refactored implementation that return statement is not going to cut the execution of the `handleKeydown` but the execution of the `handleTab` method, and there is the error. When refactoring and breaking down methods that has a return statement in the middle of the method, we need to ensure that we keep that behaviour after the refactor and make sure the method keeps stopping the execution in the same place it was doing it before the refactor. Right now with the current refactor in https://github.com/BabyElias/kolibri-design-system/tree/code-refactor-trial the code is:

```javascript
if (focusedElementIndex < focusableElements.length - 1) {
  focusableElements[focusedElementIndex + 1].focus();
  event.preventDefault();
  return [rowIndex, colIndex];
}
```

So, you are returning the current `rowIndex, colIndex` as a result of `handleTab`, returning this means that the given cell is the one that need to be focused when you press the tab, but that is an error, as the `handleKeydown` is suposed to stop its execution and avoid calling the `focusCell` method if that condition `focusedElementIndex < focusableElements.length - 1` is met, because we are already focusing the thing that need to be focused inside the handle tab method (the line `focusableElements[focusedElementIndex + 1].focus();`) and nothing more needs to be done after the `handleTab` method returns.

But, as we are returning `return [rowIndex, colIndex];` from `handleTab` that means that this is the cell that needs to be focused as a result of `handleKeyDown`, but its not the case as `handleKeyDown` has aready done its job after focusing the focusableElement, and we need to cut its execution there.

Soo, we can see it as `handleTab` process and returns the cell that needs to be focused when we press tab, but in this specific case we need to return something that tells the caller method that there is no cell to focus, as handleTab already took care of that. So an option is to just keep the empty return statements there inside the `handleTab` mthod that makes the return to be `undefined`(or return `null`) instead of the current `return [rowIndex, colIndex];`. So returning these falsy values will mean `there is no cell that need to be focused` and then we can cut the execution of the `handleKeydown` method there.

So, a fixed example would be:
```javascript
// inside `handleTab`
if (focusedElementIndex < focusableElements.length - 1) {
  focusableElements[focusedElementIndex + 1].focus();
  event.preventDefault();
  return null;
}
```

```javascript
// inside handleKeyDown
case 'Tab':
  const nextCell = this.handleTab(
    event,
    rowIndex,
    colIndex,
    totalRows,
    totalCols
  );
  if (!nextCell) {
    // cut the executing if there is no cell to be focused
    return;
  }
  [nextRowIndex, nextColIndex] = nextCell;
  break;
```

And repeat this for every `return;` statement that stopped the execution of the `handleKeyDown` method before the refactor :). Hope this helps :hugs:. And again its not required that you do this right away :sweat_smile:.

Reply:
Thank you @AlexVelezLl, much appreciated, we both got stuck here. @BabyElias Alex noticed what's going on, so I assured him it's still valuable to post here since it will help us to unblock it eventually even though in the scope of this PR we may not pursue refactor anymore. With that we can reference his guidance from a new issue as we log them tomorrow together.

Reply:
This has been logged to a doc from which we will open follow-ups. Marking as resolved in the scope of this PR.